enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.server.security.enterprise.auth;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Ticker;
+import org.apache.shiro.cache.Cache;
+import org.apache.shiro.cache.CacheException;
+import org.apache.shiro.cache.CacheManager;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+class ShiroCaffeineCache<K, V> implements Cache<K,V>
+{
+    private final com.github.benmanes.caffeine.cache.Cache<K,V> cacheMap;
+
+    ShiroCaffeineCache( Ticker ticker, long ttl, int maxCapacity )
+    {
+        if ( ttl <= 0 )
+        {
+            throw new IllegalArgumentException( "TTL must be larger than zero." );
+        }
+        cacheMap = Caffeine.newBuilder()
+                .maximumSize( maxCapacity )
+                .expireAfterWrite( ttl, TimeUnit.MILLISECONDS )
+                .ticker( ticker )
+                .build();
+    }
+
+    @Override
+    public V get( K key ) throws CacheException
+    {
+        return cacheMap.getIfPresent( key );
+    }
+
+    @Override
+    public V put( K key, V value ) throws CacheException
+    {
+        V oldValue = cacheMap.getIfPresent( key );
+        cacheMap.put( key, value );
+        return oldValue;
```

Comment:
We considered this, but it's really what we need. That get does `if cached, return; otherwise create, cache and return`while we need `if cached, return old and cache new; otherwise cache new and return null`.